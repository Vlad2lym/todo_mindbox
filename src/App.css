/* .wrapper {
  margin: 0 auto;
  width: 600px;
  display: flex;
  justify-content: center;
  align-items: center;
} */