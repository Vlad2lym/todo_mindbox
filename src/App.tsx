import React, { useEffect, useState } from 'react';
import './App.css';
import TodoList from './TodoComponents/TodoList';
import { ITodo } from './types/types';
import Context, { AppContextInterface } from './context';
import AddTodo from './TodoComponents/AddTodo';

function App() {
  const [todos, setTodos] = useState<ITodo[]> ([])
  const [countActiveTodo, setCountActiveTodo] = useState <number>(0)

  useEffect(()=> {
    setCountActiveTodo (todos.reduce((sum, todo) => {if(todo.active) sum++; return sum}, 0))
  }, [todos])

  const changeTodo: AppContextInterface = {

     removeTodo: (id: number) => {
      setTodos (todos.filter(todo => todo.id !== id))
    },

    toggleTodo: (id: number) => {
      setTodos (
        todos.map (todo => {
          if (todo.id === id) {todo.active = !todo.active}
          return todo
        })
      )
    }
  }

  const addTodo = (title: string) => {
    setTodos (todos.concat([{
      id: Date.now(),
      value: title,
      active: true
    }]))
  }

  const removeCompleted = () => {
    setTodos (todos.filter(todo=> todo.active))
  }

  return (
    <Context.Provider value={changeTodo}>
      <div className='wrapper'>
        <h1 className='title'>
          todos
        </h1>
        <AddTodo onCreate={addTodo} />
        {todos.length
          ?
          <TodoList todos={todos} clearCompleted={removeCompleted} countActive={countActiveTodo}/>
          :
          <h2 className='noTodo'>there is no todos</h2>
        }
      </div>
    </Context.Provider>
  );
}

export default App;
