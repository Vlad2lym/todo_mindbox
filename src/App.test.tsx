import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import App from './App';

describe('test app', () => {
  test('basic html', () => {
    render(<App />);
    const title = screen.getByText('todos');
    const btn = screen.getByRole('button');
    const input = screen.getByPlaceholderText(/what\sneeds\sto\sbe\sdone./i)
    expect(title).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  })

  test('input and button test', async () => {
    render(<App />);
    const noTodoTitle = screen.getByText(/is\sno\stodos/i)
    expect(noTodoTitle).toBeInTheDocument()
    const input = screen.getByPlaceholderText(/what\sneeds\sto\sbe\sdone./i)
    fireEvent.change(input, {target: {value: 'test'}})
    const btnAdd = screen.getByRole('button');
    fireEvent.click(btnAdd)
    expect(screen.queryByText(/is\sno\stodos/i)).toBeNull()
    const todoList = screen.getByTestId('todoList')
    expect(todoList).toMatchSnapshot()
    //button test
    const btnAll = screen.getByTestId('btnAll')
    const btnActive = screen.getByTestId('btnAct')
    const btnCompleted = screen.getByTestId('btnComp')
    fireEvent.click(btnCompleted)
    expect(screen.getByText(/no\scompleted\stodos$/i)).toBeInTheDocument()
    expect(screen.getByTestId('todoList')).toMatchSnapshot()
    fireEvent.click(btnActive)
    expect(screen.queryByText(/no\scompleted\stodos$/i)).toBeNull()
    const checkbox = screen.getByTestId('checkbox')
    fireEvent.click(checkbox)
    expect(screen.getByText(/no\sactive\stodos$/i)).toBeInTheDocument()
    expect(screen.getByTestId('todoList')).toMatchSnapshot()
    fireEvent.click(btnAll)
    expect(screen.queryByText(/no\sactive\stodos$/i)).toBeNull()
    const btnClearCompleted = screen.getByTestId('btnClearComp')
    expect(screen.queryByText(/is\sno\stodos/i)).toBeNull()
    fireEvent.click(btnClearCompleted)
    expect(screen.getByText(/is\sno\stodos/i)).toBeInTheDocument()
  })
})

