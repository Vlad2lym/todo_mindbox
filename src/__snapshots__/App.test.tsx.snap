// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test app basic html 1`] = `
<input
  class="inputAdd"
  placeholder=" What needs to be done?"
  type="text"
  value=""
/>
`;

exports[`test app input and button test 1`] = `
<div
  class="list"
  data-testid="todoList"
>
  <div
    class="item"
  >
    <span
      class="todoText"
    >
      <input
        class="checkbox"
        data-testid="checkbox"
        type="checkbox"
      />
      test
    </span>
    <button>
      Ã—
    </button>
  </div>
  <div
    class="footer"
  >
    <p>
      1
       items left
    </p>
    <div>
      <button
        class="active"
        data-testid="btnAll"
      >
        ALL
      </button>
      <button
        class=""
        data-testid="btnAct"
      >
        Active
      </button>
      <button
        class=""
        data-testid="btnComp"
      >
        Completed
      </button>
    </div>
    <button
      data-testid="btnClearComp"
    >
      Clear completed
    </button>
  </div>
</div>
`;

exports[`test app input and button test 2`] = `
<div
  class="list"
  data-testid="todoList"
>
  <h1
    class="noTodo"
  >
    no completed todos
  </h1>
  <div
    class="footer"
  >
    <p>
      1
       items left
    </p>
    <div>
      <button
        class=""
        data-testid="btnAll"
      >
        ALL
      </button>
      <button
        class=""
        data-testid="btnAct"
      >
        Active
      </button>
      <button
        class="active"
        data-testid="btnComp"
      >
        Completed
      </button>
    </div>
    <button
      data-testid="btnClearComp"
    >
      Clear completed
    </button>
  </div>
</div>
`;

exports[`test app input and button test 3`] = `
<div
  class="list"
  data-testid="todoList"
>
  <h1
    class="noTodo"
  >
    no active todos
  </h1>
  <div
    class="footer"
  >
    <p>
      0
       items left
    </p>
    <div>
      <button
        class=""
        data-testid="btnAll"
      >
        ALL
      </button>
      <button
        class="active"
        data-testid="btnAct"
      >
        Active
      </button>
      <button
        class=""
        data-testid="btnComp"
      >
        Completed
      </button>
    </div>
    <button
      data-testid="btnClearComp"
    >
      Clear completed
    </button>
  </div>
</div>
`;
